### PR TITLE
Payements | Gift

### PR DESCRIPTION
**Pull request by range-kun**
- Changed name edit_change_balance to accept_balance_change
- Added ItemPurchaseStatusChanger to validate income request for ItemsPurchase update through view
- Changed name gift_item_to_other_user to cancel_gift_item and added logic
- get_item_for_self_user task added vcalidations
- RefundData updated to ItemPurchaseData
- Into TransferHistory added new field account_from with GenericForeignKey
- Refund.py in item_purchase deleted and added new file item_purchase_completer with ItemPurchaseCompleter to finish process for item_purchase models
- updated invoice_execution.py, all db operations execute in transaction inside execute_invoice_operations
- Added MultipleObjectsReturned into ItemPurchaseCompleter to handel scenario where it's several ItemPurchases
- Added assert_never to handel not implemented ItemPurchase.ItemPurchaseStatus

### CODE REVIEW

File: backend/payments/apps/item_purchases/views.py [hunk 1]
```diff
@@ -38,28 +36,13 @@ def create(self, request, *args, **kwargs):
         return Response({'response': response}, status=status.HTTP_200_OK)
 
 
-class RefundView(viewsets.ViewSet, DRFtoDataClassMixin):
+class ItemPurchaseUpdaterViewSet(viewsets.ViewSet, ItemPurchaseStatusChanger):
```
Comment (lsoulworkerl) [inline backend/payments/apps/item_purchases/views.py:39]: Maybe rename it to ItemPurchaseUpdateView?

File: backend/payments/apps/base/classes.py [hunk 2]
```diff
@@ -51,3 +60,69 @@ def __convert_drf_to_dataclass(self, data_model, serializer: dict):
...
+                # PAID because this function should be called only when it's gift
+                item_purchase_processor.accept_gift()
+                return Response(status=status.HTTP_200_OK)
+            case ItemPurchase.ItemPurchaseStatus.REFUNDED:
+                item_purchase_processor.take_refund()
+                return Response(status=status.HTTP_202_ACCEPTED)
+            case _ as not_implemented:
+                assert_never(not_implemented)
+
+    def fetch_item_purchase(self, item_purchase_data: ItemPurchaseData) -> ItemPurchase | Response:
```
Comment (lsoulworkerl) [inline backend/payments/apps/base/classes.py:103]: Maybe we need another error if we don't have this user?

File: backend/payments/apps/base/classes.py [hunk 3]
```diff
@@ -51,3 +60,69 @@ def __convert_drf_to_dataclass(self, data_model, serializer: dict):
...
+        except DifferentStructureError:
+            return Response(status=status.HTTP_500_INTERNAL_SERVER_ERROR)
+
+        response = self.fetch_item_purchase(item_purchase_data)
+        if type(response) != ItemPurchase:
+            return response
+
+        data = {'item_purchase': response}
+        if event_type == ItemPurchase.ItemPurchaseStatus.PAID:
+            # PAID because this function should be called only when it's gift
```
Comment (lsoulworkerl) [inline backend/payments/apps/base/classes.py:82]: Maybe don't use these comments write on in the start of method, like:
"""
Status PAID - using only when it's gift
Status PENDING - using only when it's purchase
"""

These types of comment it's easier to see and the reader is immediately ready to read the code with these clarifications.

File: backend/payments/apps/base/classes.py [hunk 4]
```diff
@@ -51,3 +60,69 @@ def __convert_drf_to_dataclass(self, data_model, serializer: dict):
             )
             raise DifferentStructureError
         return data_model_data
+
+
+class ItemPurchaseStatusChanger(DRFtoDataClassMixin):
```
Comment (lsoulworkerl) [inline backend/payments/apps/base/classes.py:65]: Maybe better place for this class is item_purchase/services folder?

File: backend/payments/apps/base/classes.py [hunk 5]
```diff
@@ -51,3 +60,69 @@ def __convert_drf_to_dataclass(self, data_model, serializer: dict):
...
+                return Response(status=status.HTTP_200_OK)
+            case ItemPurchase.ItemPurchaseStatus.REFUNDED:
+                item_purchase_processor.take_refund()
+                return Response(status=status.HTTP_202_ACCEPTED)
+            case _ as not_implemented:
+                assert_never(not_implemented)
+
+    def fetch_item_purchase(self, item_purchase_data: ItemPurchaseData) -> ItemPurchase | Response:
+        try:
+            item_purchase = ItemPurchase.objects.get(
```
Comment (lsoulworkerl) [inline backend/payments/apps/base/classes.py:105]: But several users can at the same time gift to the user the same game. I think to resolve it we need to add not only account_to and add account_from.

File: backend/payments/apps/item_purchases/models.py [hunk 6]
```diff
@@ -16,9 +16,20 @@
 
 
 class TransferHistory(models.Model):
```
Comment (lsoulworkerl) [inline backend/payments/apps/item_purchases/models.py:18]: Add comment in the start that account_to or account_from can be Account model or Owner model.

File: backend/payments/apps/item_purchases/models.py [hunk 7]
```diff
@@ -16,9 +16,20 @@
...
+        on_delete=models.CASCADE,
+        related_name='transfer_history_account_from',
+    )
+    account_to_type = models.ForeignKey(
+        ContentType,
+        on_delete=models.CASCADE,
+        related_name='transfer_history_account_to',
+    )
+
     object_id = models.PositiveIntegerField()
```
Comment (lsoulworkerl) [inline backend/payments/apps/item_purchases/models.py:30]: Will it work fine when we use one object_id to different fields?
Example:
Transaction from Owner (1) to Account(2)

U need to transfer money from Owner with id 1 and Account with id 2, they have different object_id so they need different fields for work fine.

File: backend/payments/apps/base/classes.py [hunk 8]
```diff
@@ -51,3 +60,69 @@ def __convert_drf_to_dataclass(self, data_model, serializer: dict):
...
+            data['is_gift'] = True
+
+        try:
+            item_purchase_processor = ItemPurchaseCompleter(**data)
+        except RefundNotAllowedError as error:
+            return Response(
+                {'detail': str(error)},
+                status=status.HTTP_400_BAD_REQUEST,
+            )
+        match event_type:
```
Comment (lsoulworkerl) [inline backend/payments/apps/base/classes.py:92]: Maybe design case for more readability like this:
```
case:
    item
    return something

case:
    item
    return something

**Comment by lsoulworkerl** [comment]
And you have mistake in name of PR, pls edit it to `Payments` )))))))

File: backend/payments/apps/base/classes.py (hunk 8 above)
Comment (range-kun) [reply backend/payments/apps/base/classes.py:92]: I really don't get it what you mean by this. This is the way how match/case works, maybe you can explain a little bit better?

File: backend/payments/apps/base/classes.py (hunk 2 above)
Comment (range-kun) [reply backend/payments/apps/base/classes.py:103]: I can do it, but I think it's really doesn't matter in this case, anyway product for such user doesn't exists and if't exists, we can get user.  We can't delete item_purchase without deleting user

Comment (lsoulworkerl) [reply backend/payments/apps/base/classes.py:103]: I agree with you

File: backend/payments/apps/base/classes.py (hunk 5 above)
Comment (lsoulworkerl) [reply backend/payments/apps/base/classes.py:105]: Yes, add it to `income_data`

File: backend/payments/apps/base/classes.py (hunk 8 above)
Comment (lsoulworkerl) [reply backend/payments/apps/base/classes.py:92]: Sorry, my bad

**Comment by lsoulworkerl** [comment]
And after changing TransferHistory model you forgot to change fixtures for this model `apps/base/fixtures/test/transfer_histories.json`

**Review by lsoulworkerl — APPROVED** [review]

**Pull request merged**